java/arcs/core/host/ParticleContext.kt

```diff
@@ -0,0 +1,212 @@
+                // call particle.onReady (handled by notify below).
+                awaitingReady.forEach { it.maybeInitiateSync() }
+            }
+        }
+    }
+
+    /**
+     * Performs the shutdown lifecycle on [particle].
+     */
+    suspend fun stopParticle() {
+        withContext(dispatcher) {
+            try {
+                particle.onShutdown()
+                particleState = ParticleState.Stopped
+            } catch (e: Exception) {
+                markParticleAsFailed(e, "onShutdown")
+            }
+        }
+    }
+
+    /**
+     * Called by [StorageProxy] (via the callback in [registerHandle]) when it receives storage
+     * events. This is responsible for driving the particle lifecycle API after startup and
+     * managing the running particle state.
+     *
+     * Write-only particles should not receive any of these events.
+     *
+     * This will be executed in the context of the StorageProxy's scheduler.
+     * TODO: error handling in event methods
+     */
+    fun notify(event: StorageEvent, handle: Handle) {
+        when (event) {
+            StorageEvent.READY -> {
+                if (awaitingReady.remove(handle) && awaitingReady.isEmpty()) {
+                    moveToReady()
+                }
+            }
+            StorageEvent.UPDATE -> {
+                if (awaitingReady.isEmpty()) {
+                    particle.onUpdate()
+                }
+            }
+            StorageEvent.DESYNC -> {
+                if (desyncedHandles.isEmpty()) {
+                    particleState = ParticleState.Desynced
+                    particle.onDesync()
+                }
+                desyncedHandles.add(handle)
+            }
+            StorageEvent.RESYNC -> {
+                desyncedHandles.remove(handle)
+                if (desyncedHandles.isEmpty()) {
+                    particle.onResync()
+                    particleState = ParticleState.Running
+                }
+            }
+        }
+    }
+
+    private fun moveToReady() {
+        particle.onReady()
+        particleState = ParticleState.Running
+        notifyReady?.invoke(particle)
+        notifyReady = null
+    }
+
+    private fun markParticleAsFailed(error: Exception, eventName: String): Exception {
+        log.error(error) { "Failure in particle ${planParticle.particleName}.$eventName()" }
+
+        if (particleState != ParticleState.MaxFailed) {
+            particleState = if (particleState.hasBeenStarted) {
+                ParticleState.Failed
+            } else {
+                ParticleState.Failed_NeverStarted
+            }
+
+            consecutiveFailureCount++
+            if (consecutiveFailureCount > MAX_CONSECUTIVE_FAILURES) {
+                particleState = ParticleState.MaxFailed
+            }
+        }
+        return error
+    }
+
+    companion object {
+        private val log = TaggedLog { "ParticleContext" }
```

Comment:
the tag could include info from the Plan and current state which might help debugging (e.g. ParticleName and State)